client/settings/phone-input/index.tsx

```diff
@@ -0,0 +1,125 @@
+/**
+ * External dependencies
+ */
+import React, { useEffect, useState, useRef } from 'react';
+import { __ } from '@wordpress/i18n';
+import intlTelInput from 'intl-tel-input';
+import './style.scss';
+
+/**
+ * Internal dependencies
+ */
+import utils from 'iti/utils';
+
+interface PhoneNumberInputProps {
+	value: string;
+	onValidationChange: ( isValid: boolean ) => void;
+	onValueChange: ( value: string ) => void;
+	inputProps: {
+		label: string;
+		ariaLabel: string;
+		name: string;
+	};
+}
+
+const PhoneNumberInput = ( {
+	onValueChange,
+	value,
+	onValidationChange = ( validation ) => validation,
+	inputProps = {
+		label: '',
+		ariaLabel: '',
+		name: '',
+	},
+}: PhoneNumberInputProps ): JSX.Element => {
+	const [
+		inputInstance,
+		setInputInstance,
+	] = useState< intlTelInput.Plugin | null >( null );
+	const inputRef = useRef< HTMLInputElement >( null );
+
+	const handlePhoneNumberInputChange = () => {
+		if ( inputInstance ) {
+			onValueChange( inputInstance.getNumber() );
+			onValidationChange( inputInstance.isValidNumber() );
+		}
+	};
+
+	const removeInternationalPrefix = ( phone: string ) => {
+		if ( inputInstance ) {
+			return phone.replace(
+				'+' + inputInstance.getSelectedCountryData().dialCode,
+				''
+			);
+		}
+
+		return phone;
+	};
+
+	useEffect( () => {
+		let iti: intlTelInput.Plugin | null = null;
+		const currentRef = inputRef.current;
+
+		const handleCountryChange = () => {
```

Comment:
I think I found a bug related to the country change. Steps to reproduce:

1. Go to Payments > Card Readers > Receipt details
2. Remove some characters from the input component so you see the invalid phone error
3. Click on the country selector and change the country
4. Observe the issue

**Actual behavior**
When the country changes the phone number is rendered incorrectly. It includes the phone prefix as part of the number. 
There is a warning thrown to the browser console:
`react_devtools_backend.js:4026 Warning: Maximum update depth exceeded. This can happen when a component calls setState inside useEffect, but useEffect either doesn't have a dependency array, or one of the dependencies changes on every render.
 `
In some ocassions I noticed that the page becomes unresponsive and eventually crashes. Unfortunately I could not find the exact combination of steps to reproduce, but I think it can be reproduced playing with the component (edit phone number and change country several times)

I recorded this [video](https://d.pr/v/RlpfR5) that shows the issue in action. Please let me know if I can give any further details

Reply:
Aha, weird bug. Thank you for pointing it out!

That doesn't happen on the Checkout page, so I believe it's something specific to the Card Readers page.

Looking...

Reply:
Hey @elazzabi I think the problem could be related to the useEffect dependency array. Now we have:

```
useEffect( () => {
		// ...
	}, [ onValueChange, onValidationChange ] );
```
Since both dependencies are functions I suspect the reference changes on every render, causing the useEffect to run again. I used a different approach that seems to work, maybe you'd like to give it a try:

```
useEffect( () => {
		// ...
	}, [ inputRef ] );
```
With this approach, the effect will run if inputRef changes which I think is correct.

Reply:
Thank you for the pointers @mgascam! I took a look and, changing the dependencies of `useEffect` to `inputRef` fixes it. However, that fails the lint process given `useEffect` uses both `onValueChange` and `onValidationChange`, and thus, they need to appear there.

So, I left them both but looked deeper into what was causing the issue. It appears that `onValidationChange` is changing on every render, and thus makes `useEffect` re-run on every render.

I wrapped a few functions with `useCallback` to avoid the issue. Now it passes, but the new problem is the input loses focus on every key stroke 😅

Reply:
Ok, back to this.

The problem with this is that the input gets re-rendered on each keystroke. This is derived from a long chain of changes:

Changes to the input are tracked via the `accountBusinessSupportPhone` variable. Any change to this variable will trigger a re-calculation of the updateSaveButtonAvailability [function](https://github.com/Automattic/woocommerce-payments/pull/4355/commits/a8b78b0adc91c15a6d5d85cb65f30799e1a83826#diff-606705683552d06907190d85b12172ed4482010bf9f80f5c39ebc94636458055R67). Each change to this function will trigger a re-calculation of the `onValueChange` [prop](https://github.com/Automattic/woocommerce-payments/pull/4355/commits/a8b78b0adc91c15a6d5d85cb65f30799e1a83826#diff-606705683552d06907190d85b12172ed4482010bf9f80f5c39ebc94636458055R107). And finally, each change to the prop will re-trigger the `useEffect` in the input component. Inside the useEffect, we attach the library to the input. So every time we execute it, the library is re-attached and the focus is lost.

The real problem is I made the `onValueChange` make so many things, which is not ideal, my bad. Especially, [the part where it checks if the Save button should be enabled](https://github.com/Automattic/woocommerce-payments/pull/4355/commits/75602c6dc37f9063349da4d104e3ab99c64081b0#diff-606705683552d06907190d85b12172ed4482010bf9f80f5c39ebc94636458055L105). Now, it is fixed to only change the value. Checking for the button enable/disable state is [happening outside](https://github.com/Automattic/woocommerce-payments/pull/4355/commits/75602c6dc37f9063349da4d104e3ab99c64081b0#diff-606705683552d06907190d85b12172ed4482010bf9f80f5c39ebc94636458055R59). This has the benefit of being independent and will be used by [other inputs too](https://github.com/Automattic/woocommerce-payments/pull/4355/commits/75602c6dc37f9063349da4d104e3ab99c64081b0#r919136794).

Reply:
Nice! I am not able to reproduce the bug anymore, well done